File: jl_src/day_10.jl
import ArgParse
import Pkg
Pkg.add("DataStructures");
import DataStructures

# Build ArgParsing Table
s = ArgParse.ArgParseSettings()
@ArgParse.add_arg_table s begin
    "--input"
        help = "Filepath to input data"
        arg_type = String
        required = true
    "--verbose"
        help = "Add additional reporting"
        action = :store_true
end

# Parse arguments
parsed_args = ArgParse.parse_args(ARGS, s)

# Process files
function ProcessFile(filepath::String)::Vector{String}
    results = Vector{String}()
    open(filepath) do f
        for line in readlines(f)
            push!(results, line)
        end
    end
    results
end

abstract type Block end
struct Regular <: Block end
struct Square <: Block end
struct Curly <: Block end
struct Angle <: Block end
struct None <: Block end

char_map = Dict{Char,Block}([ 
    ('(', Regular()),
    (')', Regular()),
    ('[', Square()),
    (']', Square()),
    ('{', Curly()),
    ('}', Curly()),
    ('<', Angle()),
    ('>', Angle())]);

function get_illegal_char(line::String)::Block
    block_stack = DataStructures.Deque{Block}();
    for ch in line
        if ch in ['(', '[', '{', '<']
            cur_block = char_map[ch]
            pushfirst!(block_stack, cur_block)
        elseif ch in [')', ']', '}', '>' ]
            cur_block = char_map[ch]
            last_block = popfirst!(block_stack)
            if cur_block != last_block
                return cur_block
            end
        else
            throw(InvalidStateException("Encountered unexpected character! ", ch));
        end
    end
    None()
end

function get_line_completion(line::String)::Vector{Block}
    result = Vector{Block}()
    block_stack = DataStructures.Deque{Block}();
    for ch in line
        if ch in ['(', '[', '{', '<']
            cur_block = char_map[ch]
            pushfirst!(block_stack, cur_block)
        elseif ch in [')', ']', '}', '>' ]
            cur_block = char_map[ch]
            last_block = popfirst!(block_stack)
            if cur_block != last_block
                return Vector{Block}([None()])
            end
        else
            throw(InvalidStateException("Encountered unexpected character! ", ch));
        end
    end
    for b in block_stack
        push!(result, b)
    end
    result
end

function score_line(blocks::Vector{Block})::UInt
    add_map = Dict{Block,UInt}([
        (Regular(), 1),
        (Square(), 2),
        (Curly(), 3),
        (Angle(), 4)
    ]);

    result = 0;
    for b in blocks
        result = result*5+add_map[b]
    end
    result
end

# Get input filepath
input_filepath = parsed_args["input"]
verbose = parsed_args["verbose"]

nav_data = ProcessFile(input_filepath)

score_map = Dict{Block,UInt}([
    (Regular(), 3),
    (Square(), 57),
    (Curly(), 1197),
    (Angle(), 25137)
])

score_blocks = filter(r -> r != None(), map(s -> get_illegal_char(s), nav_data))
syntax_error_score = sum(map(b -> score_map[b], score_blocks))
println("Day 10 problem 1: ", syntax_error_score)

function v_test(v::Vector{Block})::Bool
    if length(v) == 1
        if v[1] == None()
            return false
        else
            return true
        end
    else
        return true
    end
end

line_blocks = filter(r -> v_test(r), map(s -> get_line_completion(s), nav_data))
line_scores = collect(map(v -> score_line(v), line_blocks))

sort!(line_scores)

middle_score = line_scores[(length(line_scores)÷2)+1];

println("Day 10 problem 2: ", middle_score);

File: rust_src/src/day_10.rs
extern crate argparse;

use std::io::{self, BufRead};
use std::fs::File;
use std::path::Path;
use std::collections::{HashMap,VecDeque};
use std::hash::Hash;
use std::iter::FromIterator;
use argparse::{ArgumentParser, StoreTrue, Store};

// Define readlines function discussed here:
// https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn process_input<P>(filename: P) -> Vec<String>
where P: AsRef<Path>, {
    let mut results = Vec::new();
    // Reading the file
    if let Ok(lines) = read_lines(filename) {
        // Consumes the iterator, returns an (Optional) String
        for line in lines {
            if let Ok(ip) = line {
                results.push(ip)
            }
        }
    } else {
		panic!("Problem opening file!");
	}
    results
}

#[derive(Debug,PartialEq,Eq,Hash,Clone,Copy)]
enum Block {
    Regular,
    Square,
    Curly,
    Angle,
}

fn get_illegal_char(line: &String) -> Option<Block> {
    let char_map: HashMap<char,Block> = HashMap::from_iter([
        ('(', Block::Regular),
        (')', Block::Regular),
        ('[', Block::Square),
        (']', Block::Square),
        ('{', Block::Curly),
        ('}', Block::Curly),
        ('<', Block::Angle),
        ('>', Block::Angle),
    ]);
    let mut block_stack: VecDeque<Block> = VecDeque::new();
    for ch in line.chars() {
        if ['(', '[', '{', '<'].contains(&ch) {
            let cur_block = char_map.get(&ch).unwrap();
            block_stack.push_front(*cur_block);
        } else if [')', ']', '}', '>' ].contains(&ch) {
            let cur_block = char_map.get(&ch).unwrap();
            let last_block = block_stack.pop_front().unwrap();
            if *cur_block != last_block {
                return Some(*cur_block);
            }
        } else {
            panic!("Encountered unexpected character! {}", ch);
        }
    }
    None
}

fn get_line_completion(line: &String) -> Option<Vec<Block>> {
    let char_map: HashMap<char,Block> = HashMap::from_iter([
        ('(', Block::Regular),
        (')', Block::Regular),
        ('[', Block::Square),
        (']', Block::Square),
        ('{', Block::Curly),
        ('}', Block::Curly),
        ('<', Block::Angle),
        ('>', Block::Angle),
    ]);
    let mut block_stack: VecDeque<Block> = VecDeque::new();
    for ch in line.chars() {
        if ['(', '[', '{', '<'].contains(&ch) {
            let cur_block = char_map.get(&ch).unwrap();
            block_stack.push_front(*cur_block);
        } else if [')', ']', '}', '>' ].contains(&ch) {
            let cur_block = char_map.get(&ch).unwrap();
            let last_block = block_stack.pop_front().unwrap();
            if *cur_block != last_block {
                return None;
            }
        } else {
            panic!("Encountered unexpected character! {}", ch);
        }
    }
    Some(block_stack.iter().copied().collect::<Vec::<Block>>())
}

fn score_line(blocks: &Vec<Block>) -> usize {
    let add_map: HashMap<Block,usize> = HashMap::from_iter([
        (Block::Regular, 1),
        (Block::Square, 2),
        (Block::Curly, 3),
        (Block::Angle, 4)
    ]);

    let mut result = 0;
    for b in blocks {
        result = result*5+add_map.get(&b).unwrap();
    }
    result
}

fn main() {
    // Argument Parsing
    let mut verbose = false;
    let mut input = String::new();
    {
        let mut ap = ArgumentParser::new();
        ap.set_description("Day 10 AOC 2021");
        ap.refer(&mut input)
            .add_option(&["-i", "--input"], Store,
            "Filepath to input data");
        ap.refer(&mut verbose)
            .add_option(&["-v", "--verbose"], StoreTrue,
            "Add additional reporting");
        ap.parse_args_or_exit();
    }

    let score_map: HashMap<Block,usize> = HashMap::from_iter([
        (Block::Regular, 3),
        (Block::Square, 57),
        (Block::Curly, 1197),
        (Block::Angle, 25137)
    ]);

    let nav_data = process_input(input);

    let syntax_error_score: usize = nav_data.iter()
        .map(|s| get_illegal_char(s))
        .filter(|r| {
            match r {
                None => false,
                Some(_) => true,
            }
        })
        .map(|r| r.unwrap())
        .map(|r| score_map.get(&r).unwrap()).sum();

    println!("Day 10 problem 1: {}", syntax_error_score);

    let mut line_scores: Vec<usize> = nav_data.iter()
        .map(|s| get_line_completion(s))
        .filter(|r| {
            match r {
                None => false,
                Some(_) => true,
            }
        })
        .map(|r| r.unwrap())
        .map(|v| {
            score_line(&v)
        }).collect::<Vec::<usize>>();

    line_scores.sort();

    let middle_score = line_scores[(line_scores.len()/2)];

    println!("Day 10 problem 2: {}", middle_score);
}
